scripts/windows/setup/IotEdgeSecurityDaemon.ps1

```diff
@@ -219,36 +219,34 @@ function Initialize-IoTEdge {
     if (-not (Test-EdgeAlreadyInstalled)) {
         Write-HostRed
         Write-HostRed ('IoT Edge is not yet installed. ' + $InstallMessage)
-        return
+        throw
```

Comment:
Consider passing the failure message through throw directly instead of Write-RedHost. This way, if something is consuming the script, it can catch the exception and display is separately.

Reply:
I thought of that, but a few places have multi-line errors built using multiple calls to `Write-HostRed`, so I wasn't sure it should get collected into a single `throw`n message.

Maybe having newlines in the string will be good enough. Let me experiment.

Reply:
It looks worse, unfortunately:

```powershell
$ Write-Host foo; throw "`nC:\ProgramData\iotedge\config.yaml already exists.`nDelete it using `"Uninstall-IoTEdge -Force`" and then re-run `"Deploy-IoTEdge`" and `"Initialize-IoTEdge`"`n"

foo
C:\ProgramData\iotedge\config.yaml already exists.
Delete it using "Uninstall-IoTEdge -Force" and then re-run "Deploy-IoTEdge" and "Initialize-IoTEdge"
At line:1 char:17
+ ... e-host foo; throw "`nC:\ProgramData\iotedge\config.yaml already exist ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : OperationStopped: (
C:\ProgramData...alize-IoTEdge"
:String) [], RuntimeException
+ FullyQualifiedErrorId : 
C:\ProgramData\iotedge\config.yaml already exists.
Delete it using "Uninstall-IoTEdge -Force" and then re-run "Deploy-IoTEdge" and "Initialize-IoTEdge"
```

It swallows any extra leading or trailing newlines in the string that we currently use to separate the error message from the rest of the output for emphasis. Also, the `FullyQualifiedErrorId` bit at the end duplicates the entire multi-line error message.

Compare with:

```powershell
$ Write-Host foo; Write-HostRed; Write-HostRed 'C:\ProgramData\iotedge\config.yaml already exists.'; Write-HostRed 'Delete it using "Uninstall-IoTEdge -Force" and then re-run "Deploy-IoTEdge" and "Initialize-IoTEdge"'; Write-HostRed; throw

foo

C:\ProgramData\iotedge\config.yaml already exists.
Delete it using "Uninstall-IoTEdge -Force" and then re-run "Deploy-IoTEdge" and "Initialize-IoTEdge"

ScriptHalted
At line:1 char:233
+ ... -run "Deploy-IoTEdge" and "Initialize-IoTEdge"'; Write-HostRed; throw
+                                                                     ~~~~~
+ CategoryInfo          : OperationStopped: (:) [], RuntimeException
+ FullyQualifiedErrorId : ScriptHalted
```

---

Any wrapper should probably be showing the full script output anyway, which would include the `Write-HostRed` output. For example our E2E tests do so.

Reply:
Up to you. One more option would be to write-host the new line before you throw.